File: frontend/src/types/verification.ts
/**
 * Verification Types for Attorney Finding Verification Workflow
 *
 * Types for the verification queue UI and finding verification.
 * Matches backend Pydantic models in app/models/verification.py
 *
 * Story 8-5: Implement Verification Queue UI
 * Epic 8: Safety Layer (Guardrails, Policing, Verification)
 */

// =============================================================================
// Story 8-5: Verification Enums (Task 8.2, 8.3)
// =============================================================================

/**
 * Attorney verification decision.
 *
 * Matches backend VerificationDecision enum.
 */
export enum VerificationDecision {
  PENDING = 'pending',
  APPROVED = 'approved',
  REJECTED = 'rejected',
  FLAGGED = 'flagged',
}

/**
 * Verification requirement tier based on confidence.
 *
 * Implements ADR-004 tiered verification:
 * - OPTIONAL: Confidence > 90% (export allowed)
 * - SUGGESTED: Confidence 70-90% (export warning shown)
 * - REQUIRED: Confidence < 70% (export blocked until verified)
 */
export enum VerificationRequirement {
  OPTIONAL = 'optional',
  SUGGESTED = 'suggested',
  REQUIRED = 'required',
}

// =============================================================================
// Story 8-5: Queue Item Types (Task 8.4)
// =============================================================================

/**
 * Item in verification queue for UI display.
 *
 * Matches backend VerificationQueueItem model.
 */
export interface VerificationQueueItem {
  /** Verification UUID */
  id: string;
  /** Finding UUID (nullable if finding was deleted) */
  findingId: string | null;
  /** Finding type for filtering (citation_mismatch, timeline_anomaly, etc.) */
  findingType: string;
  /** Summary for queue display */
  findingSummary: string;
  /** Confidence percentage (0-100) */
  confidence: number;
  /** Verification requirement tier */
  requirement: VerificationRequirement;
  /** Current decision status */
  decision: VerificationDecision;
  /** When finding was created */
  createdAt: string;
  /** Primary source document name */
  sourceDocument: string | null;
  /** Source engine (citation, timeline, contradiction) */
  engine: string;
  /** User UUID who verified (AC #2 - recorded on approval) */
  verifiedBy?: string | null;
  /** When verified (AC #2 - recorded on approval) */
  verifiedAt?: string | null;
  /** Attorney notes for decision */
  notes?: string | null;
}

// =============================================================================
// Story 8-5: Statistics Types (Task 8.5)
// =============================================================================

/**
 * Verification statistics for dashboard.
 *
 * Matches backend VerificationStats model.
 */
export interface VerificationStats {
  /** Total verification records */
  totalVerifications: number;
  /** Awaiting review */
  pendingCount: number;
  /** Approved by attorney */
  approvedCount: number;
  /** Rejected by attorney */
  rejectedCount: number;
  /** Flagged for further review */
  flaggedCount: number;
  /** < 70% confidence, pending (blocks export) */
  requiredPending: number;
  /** 70-90% confidence, pending */
  suggestedPending: number;
  /** > 90% confidence, pending */
  optionalPending: number;
  /** True if has unverified findings < 70% confidence */
  exportBlocked: boolean;
  /** Count of findings blocking export */
  blockingCount: number;
}

// =============================================================================
// Story 8-5: Filter Types (Task 8.6)
// =============================================================================

/**
 * Confidence tier for filtering.
 */
export type ConfidenceTier = 'high' | 'medium' | 'low';

/**
 * View mode for verification queue.
 */
export type VerificationView = 'queue' | 'by-type' | 'history';

/**
 * Filter state for verification queue.
 */
export interface VerificationFilters {
  /** Filter by finding type */
  findingType: string | null;
  /** Filter by confidence tier (high >90%, medium 70-90%, low <70%) */
  confidenceTier: ConfidenceTier | null;
  /** Filter by verification status */
  status: VerificationDecision | null;
  /** Current view mode */
  view: VerificationView;
}

// =============================================================================
// Story 8-5: API Response Types
// =============================================================================

/**
 * Response for verification queue endpoint.
 */
export interface VerificationQueueResponse {
  data: VerificationQueueItem[];
  meta: {
    limit: number;
    count: number;
  };
}

/**
 * Response for verification stats endpoint.
 */
export interface VerificationStatsResponse {
  data: VerificationStats;
}

/**
 * Complete verification record from database.
 */
export interface FindingVerification {
  id: string;
  matterId: string;
  findingId: string | null;
  findingType: string;
  findingSummary: string;
  confidenceBefore: number;
  decision: VerificationDecision;
  verifiedBy: string | null;
  verifiedAt: string | null;
  confidenceAfter: number | null;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
  verificationRequirement: VerificationRequirement;
}

/**
 * Response for single verification record.
 */
export interface VerificationResponse {
  data: FindingVerification;
}

/**
 * Response for verification list endpoint.
 */
export interface VerificationListResponse {
  data: FindingVerification[];
}

/**
 * Response for bulk verification operations.
 */
export interface BulkVerificationResponse {
  data: Record<string, unknown>;
  updatedCount: number;
  failedIds: string[];
}

// =============================================================================
// Story 8-5: Request Types
// =============================================================================

/**
 * Request body for approving a verification.
 */
export interface ApproveVerificationRequest {
  notes?: string;
  confidenceAfter?: number;
}

/**
 * Request body for rejecting a verification.
 */
export interface RejectVerificationRequest {
  notes: string;
}

/**
 * Request body for flagging a verification.
 */
export interface FlagVerificationRequest {
  notes: string;
}

/**
 * Request for bulk verification operations.
 */
export interface BulkVerificationRequest {
  verificationIds: string[];
  decision: VerificationDecision;
  notes?: string;
}

// =============================================================================
// Story 12-3: Export Eligibility Types (Task 1)
// =============================================================================

/**
 * Finding that blocks export.
 *
 * Story 12-3: AC #1 - Details of finding blocking export.
 */
export interface ExportBlockingFinding {
  verificationId: string;
  findingId: string | null;
  findingType: string;
  findingSummary: string;
  confidence: number;
}

/**
 * Finding that shows warning but doesn't block export.
 *
 * Story 12-3: AC #2 - Findings with 70-90% confidence show warning.
 */
export interface ExportWarningFinding {
  verificationId: string;
  findingId: string | null;
  findingType: string;
  findingSummary: string;
  confidence: number;
}

/**
 * Result of export eligibility check.
 *
 * Story 12-3: AC #1, #2 - Export eligibility with blocking and warning findings.
 */
export interface ExportEligibility {
  eligible: boolean;
  blockingFindings: ExportBlockingFinding[];
  blockingCount: number;
  warningFindings: ExportWarningFinding[];
  warningCount: number;
  message: string;
}

/**
 * Response for export eligibility endpoint.
 */
export interface ExportEligibilityResponse {
  data: ExportEligibility;
}


// =============================================================================
// Verification Priority (Lawyer UX Improvement)
// =============================================================================

/**
 * Priority level for verification items.
 * Used to help lawyers focus on most urgent items first.
 */
export type VerificationPriority = 'urgent' | 'high' | 'normal' | 'low';

/**
 * Priority analysis result for a verification item.
 */
export interface VerificationPriorityAnalysis {
  /** Calculated priority level */
  priority: VerificationPriority;
  /** Priority score (higher = more urgent) */
  score: number;
  /** Reasons for this priority */
  reasons: string[];
  /** Badge color class for display */
  badgeClass: string;
  /** Label for display */
  label: string;
}

/**
 * Calculate priority for a verification item.
 *
 * Factors considered:
 * - Requirement level (REQUIRED > SUGGESTED > OPTIONAL)
 * - Confidence score (lower = higher priority)
 * - Age (older items get slight boost)
 * - Finding type (contradictions often need faster review)
 *
 * @example
 * ```ts
 * const analysis = calculateVerificationPriority(item);
 * // { priority: 'urgent', label: 'For Immediate Review', reasons: [...] }
 * ```
 */
export function calculateVerificationPriority(
  item: VerificationQueueItem
): VerificationPriorityAnalysis {
  let score = 0;
  const reasons: string[] = [];

  // Requirement level weight (most important)
  switch (item.requirement) {
    case VerificationRequirement.REQUIRED:
      score += 50;
      reasons.push('Required verification (blocks export)');
      break;
    case VerificationRequirement.SUGGESTED:
      score += 30;
      reasons.push('Suggested verification');
      break;
    case VerificationRequirement.OPTIONAL:
      score += 10;
      break;
  }

  // Confidence score weight (lower confidence = higher priority)
  if (item.confidence < 50) {
    score += 30;
    reasons.push('Very low confidence (<50%)');
  } else if (item.confidence < 70) {
    score += 20;
    reasons.push('Low confidence (<70%)');
  } else if (item.confidence < 85) {
    score += 10;
  }

  // Finding type weight (some findings need faster review)
  if (item.findingType === 'contradiction' || item.findingType === 'factual_mismatch') {
    score += 15;
    reasons.push('Contradiction finding');
  } else if (item.findingType === 'citation_mismatch') {
    score += 10;
    reasons.push('Citation verification needed');
  }

  // Age weight (items older than 7 days get boost)
  const createdDate = new Date(item.createdAt);
  const daysSinceCreated = Math.floor(
    (Date.now() - createdDate.getTime()) / (1000 * 60 * 60 * 24)
  );
  if (daysSinceCreated > 14) {
    score += 15;
    reasons.push('Pending for over 2 weeks');
  } else if (daysSinceCreated > 7) {
    score += 10;
    reasons.push('Pending for over a week');
  }

  // Determine priority level from score
  let priority: VerificationPriority;
  let label: string;
  let badgeClass: string;

  if (score >= 70) {
    priority = 'urgent';
    label = 'For Immediate Review';
    badgeClass = 'bg-red-100 text-red-800 border-red-200 dark:bg-red-900/30 dark:text-red-400';
  } else if (score >= 50) {
    priority = 'high';
    label = 'Review Soon';
    badgeClass = 'bg-orange-100 text-orange-800 border-orange-200 dark:bg-orange-900/30 dark:text-orange-400';
  } else if (score >= 30) {
    priority = 'normal';
    label = 'Standard Review';
    badgeClass = 'bg-blue-100 text-blue-800 border-blue-200 dark:bg-blue-900/30 dark:text-blue-400';
  } else {
    priority = 'low';
    label = 'Can Wait';
    badgeClass = 'bg-gray-100 text-gray-600 border-gray-200 dark:bg-gray-800 dark:text-gray-400';
  }

  return {
    priority,
    score,
    reasons,
    badgeClass,
    label,
  };
}

/**
 * Sort verification items by priority (highest first).
 */
export function sortByPriority(items: VerificationQueueItem[]): VerificationQueueItem[] {
  return [...items].sort((a, b) => {
    const priorityA = calculateVerificationPriority(a);
    const priorityB = calculateVerificationPriority(b);
    return priorityB.score - priorityA.score;
  });
}

File: backend/app/models/verification.py
"""Verification models for attorney finding verification workflow.

Story 8-4: Implement Finding Verifications Table
Epic 8: Safety Layer (Guardrails, Policing, Verification)

These models define the structure for the attorney verification workflow:
- VerificationDecision: Attorney decision enum (pending, approved, rejected, flagged)
- VerificationRequirement: Tier enum based on confidence (optional, suggested, required)
- FindingVerificationCreate: Model for creating new verification records
- FindingVerificationUpdate: Model for updating verification decisions
- FindingVerification: Complete verification record from database
- VerificationQueueItem: Optimized model for queue UI display
- VerificationStats: Aggregate statistics for dashboard

Implements:
- FR10: Attorney Verification Workflow
- NFR23: Court-defensible verification workflow with forensic trail
- ADR-004: Verification Tier Thresholds (>90% optional, 70-90% suggested, <70% required)
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

# =============================================================================
# Story 8-4: Verification Enums (Task 3.2, 3.3)
# =============================================================================


class VerificationDecision(str, Enum):
    """Attorney verification decision.

    Story 8-4: AC #2 - Decision options for finding verification.
    """

    PENDING = "pending"  # Awaiting attorney review
    APPROVED = "approved"  # Attorney verified finding is correct
    REJECTED = "rejected"  # Attorney marked finding as incorrect
    FLAGGED = "flagged"  # Attorney flagged for further review


class VerificationRequirement(str, Enum):
    """Verification requirement based on confidence.

    Story 8-4: AC #3-5 - Implements ADR-004 tiered verification.

    Thresholds:
    - OPTIONAL: Confidence > 90% (export allowed, verification informational)
    - SUGGESTED: Confidence 70-90% (export warning shown, verification suggested)
    - REQUIRED: Confidence < 70% (export blocked until verified)
    """

    OPTIONAL = "optional"  # Confidence > 90%
    SUGGESTED = "suggested"  # Confidence 70-90%
    REQUIRED = "required"  # Confidence < 70%


# =============================================================================
# Story 8-4: Verification Create/Update Models (Task 3.4, 3.5)
# =============================================================================


class FindingVerificationCreate(BaseModel):
    """Create verification record when finding is generated.

    Story 8-4: AC #1 - Created automatically when finding generated.
    Called by ResultAggregator after each finding is created.
    """

    model_config = ConfigDict(populate_by_name=True)

    matter_id: str = Field(..., alias="matterId", description="Matter UUID")
    finding_id: str = Field(..., alias="findingId", description="Finding UUID")
    finding_type: str = Field(
        ...,
        alias="findingType",
        description="Type: citation_mismatch, timeline_anomaly, contradiction, etc.",
    )
    finding_summary: str = Field(
        ...,
        max_length=500,
        alias="findingSummary",
        description="Brief description for queue display (truncated to 500 chars)",
    )
    confidence_before: float = Field(
        ...,
        ge=0,
        le=100,
        alias="confidenceBefore",
        description="Confidence at creation (0-100 scale)",
    )


class FindingVerificationUpdate(BaseModel):
    """Update verification with attorney decision.

    Story 8-4: AC #2 - Attorney approval/rejection.
    """

    model_config = ConfigDict(populate_by_name=True)

    decision: VerificationDecision = Field(..., description="Attorney decision")
    confidence_after: float | None = Field(
        None,
        ge=0,
        le=100,
        alias="confidenceAfter",
        description="Attorney-adjusted confidence (optional)",
    )
    notes: str | None = Field(
        None,
        max_length=2000,
        description="Attorney notes explaining decision",
    )


# =============================================================================
# Story 8-4: Complete Verification Model (Task 3.6)
# =============================================================================


class FindingVerification(BaseModel):
    """Complete verification record from database.

    Story 8-4: Full verification record with all fields.
    """

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., description="Verification UUID")
    matter_id: str = Field(..., alias="matterId", description="Matter UUID")
    finding_id: str | None = Field(
        None,
        alias="findingId",
        description="Finding UUID (nullable if finding was deleted)",
    )
    finding_type: str = Field(..., alias="findingType", description="Finding type snapshot")
    finding_summary: str = Field(..., alias="findingSummary", description="Finding summary snapshot")
    confidence_before: float = Field(
        ...,
        ge=0,
        le=100,
        alias="confidenceBefore",
        description="Original confidence at finding creation",
    )
    decision: VerificationDecision = Field(..., description="Current verification decision")
    verified_by: str | None = Field(None, alias="verifiedBy", description="Verifier user UUID")
    verified_at: datetime | None = Field(None, alias="verifiedAt", description="Verification timestamp")
    confidence_after: float | None = Field(
        None,
        ge=0,
        le=100,
        alias="confidenceAfter",
        description="Attorney-adjusted confidence",
    )
    notes: str | None = Field(None, description="Attorney notes")
    created_at: datetime = Field(..., alias="createdAt", description="Record creation timestamp")
    updated_at: datetime = Field(..., alias="updatedAt", description="Last update timestamp")

    # Computed field - set by service based on confidence_before
    verification_requirement: VerificationRequirement = Field(
        ...,
        alias="verificationRequirement",
        description="OPTIONAL/SUGGESTED/REQUIRED based on confidence (ADR-004)",
    )


# =============================================================================
# Story 8-4: Queue and Stats Models (Task 3.7, 3.8)
# =============================================================================


class VerificationQueueItem(BaseModel):
    """Item in verification queue for UI.

    Story 8-4: Optimized for Story 8-5 queue display.
    Includes finding context for attorney prioritization.
    """

    model_config = ConfigDict(populate_by_name=True)

    id: str = Field(..., description="Verification UUID")
    finding_id: str | None = Field(None, alias="findingId", description="Finding UUID")
    finding_type: str = Field(..., alias="findingType", description="Finding type for filtering")
    finding_summary: str = Field(..., alias="findingSummary", description="Summary for queue display")
    confidence: float = Field(
        ...,
        ge=0,
        le=100,
        description="Confidence percentage",
    )
    requirement: VerificationRequirement = Field(
        ...,
        description="Verification requirement tier",
    )
    decision: VerificationDecision = Field(..., description="Current decision status")
    created_at: datetime = Field(..., alias="createdAt", description="When finding was created")

    # Context for queue prioritization
    source_document: str | None = Field(
        None,
        alias="sourceDocument",
        description="Primary source document name",
    )
    engine: str = Field(
        ...,
        description="Source engine (citation, timeline, contradiction)",
    )


class VerificationStats(BaseModel):
    """Verification statistics for dashboard.

    Story 8-4: Aggregate stats for matter verification status.
    Used by dashboard to show verification progress.
    """

    model_config = ConfigDict(populate_by_name=True)

    total_verifications: int = Field(0, ge=0, alias="totalVerifications", description="Total verification records")
    pending_count: int = Field(0, ge=0, alias="pendingCount", description="Awaiting review")
    approved_count: int = Field(0, ge=0, alias="approvedCount", description="Approved by attorney")
    rejected_count: int = Field(0, ge=0, alias="rejectedCount", description="Rejected by attorney")
    flagged_count: int = Field(0, ge=0, alias="flaggedCount", description="Flagged for further review")

    # By requirement tier (pending only)
    required_pending: int = Field(
        0,
        ge=0,
        alias="requiredPending",
        description="< 70% confidence, pending (blocks export)",
    )
    suggested_pending: int = Field(
        0,
        ge=0,
        alias="suggestedPending",
        description="70-90% confidence, pending",
    )
    optional_pending: int = Field(
        0,
        ge=0,
        alias="optionalPending",
        description="> 90% confidence, pending",
    )

    # Export eligibility
    export_blocked: bool = Field(
        False,
        alias="exportBlocked",
        description="True if has unverified findings < 70% confidence",
    )
    blocking_count: int = Field(
        0,
        ge=0,
        alias="blockingCount",
        description="Count of findings blocking export",
    )


# =============================================================================
# Story 8-4: API Response Models
# =============================================================================


class VerificationListResponse(BaseModel):
    """Response for verification list endpoint.

    Story 8-4: Follows project API response pattern with data wrapper.
    """

    data: list[FindingVerification] = Field(
        default_factory=list,
        description="List of verification records",
    )


class VerificationQueueResponse(BaseModel):
    """Response for verification queue endpoint.

    Story 8-4: Optimized for Story 8-5 queue UI.
    """

    data: list[VerificationQueueItem] = Field(
        default_factory=list,
        description="Pending verification items",
    )
    meta: dict = Field(
        default_factory=dict,
        description="Pagination and filter metadata",
    )


class VerificationStatsResponse(BaseModel):
    """Response for verification stats endpoint.

    Story 8-4: Dashboard statistics.
    """

    data: VerificationStats = Field(..., description="Verification statistics")


class VerificationResponse(BaseModel):
    """Response for single verification record.

    Story 8-4: Follows project API response pattern.
    """

    data: FindingVerification = Field(..., description="Verification record")


# =============================================================================
# Story 8-4: Action Request Models (Code Review Fix)
# =============================================================================


class ApproveVerificationRequest(BaseModel):
    """Request body for approving a verification.

    Story 8-4: Code Review Fix - Use request body instead of query params.
    """

    model_config = ConfigDict(populate_by_name=True)

    notes: str | None = Field(
        None,
        max_length=2000,
        description="Optional approval notes",
    )
    confidence_after: float | None = Field(
        None,
        ge=0,
        le=100,
        alias="confidenceAfter",
        description="Optional adjusted confidence score",
    )


class RejectVerificationRequest(BaseModel):
    """Request body for rejecting a verification.

    Story 8-4: Code Review Fix - Use request body instead of query params.
    """

    notes: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="Required rejection notes",
    )


class FlagVerificationRequest(BaseModel):
    """Request body for flagging a verification.

    Story 8-4: Code Review Fix - Use request body instead of query params.
    """

    notes: str = Field(
        ...,
        min_length=1,
        max_length=2000,
        description="Required flagging notes",
    )


class BulkVerificationRequest(BaseModel):
    """Request for bulk verification operations.

    Story 8-4: Task 4.8 - Bulk approve/reject for Story 8-5 queue UI.
    """

    model_config = ConfigDict(populate_by_name=True)

    verification_ids: list[str] = Field(
        ...,
        min_length=1,
        max_length=100,
        alias="verificationIds",
        description="List of verification UUIDs to update (max 100)",
    )
    decision: VerificationDecision = Field(
        ...,
        description="Decision to apply to all",
    )
    notes: str | None = Field(
        None,
        max_length=2000,
        description="Optional notes for all verifications",
    )


class BulkVerificationResponse(BaseModel):
    """Response for bulk verification operations.

    Story 8-4: Task 4.8 - Results of bulk operation.
    """

    model_config = ConfigDict(populate_by_name=True)

    data: dict = Field(
        default_factory=dict,
        description="Bulk operation results",
    )
    updated_count: int = Field(0, ge=0, alias="updatedCount", description="Number of records updated")
    failed_ids: list[str] = Field(
        default_factory=list,
        alias="failedIds",
        description="IDs that failed to update",
    )


# =============================================================================
# Story 8-4: Export Eligibility Models (Task 8)
# =============================================================================


class ExportBlockingFinding(BaseModel):
    """Finding that blocks export.

    Story 8-4: AC #5 - Details of finding blocking export.
    """

    model_config = ConfigDict(populate_by_name=True)

    verification_id: str = Field(..., alias="verificationId", description="Verification UUID")
    finding_id: str | None = Field(None, alias="findingId", description="Finding UUID")
    finding_type: str = Field(..., alias="findingType", description="Finding type")
    finding_summary: str = Field(..., alias="findingSummary", description="Finding summary")
    confidence: float = Field(..., ge=0, le=100, description="Confidence score")


class ExportWarningFinding(BaseModel):
    """Finding that shows warning but doesn't block export.

    Story 12-3: AC #2 - Findings with 70-90% confidence show warning.
    """

    model_config = ConfigDict(populate_by_name=True)

    verification_id: str = Field(..., alias="verificationId", description="Verification UUID")
    finding_id: str | None = Field(None, alias="findingId", description="Finding UUID")
    finding_type: str = Field(..., alias="findingType", description="Finding type")
    finding_summary: str = Field(..., alias="findingSummary", description="Finding summary")
    confidence: float = Field(..., ge=0, le=100, description="Confidence score")


class ExportEligibilityResult(BaseModel):
    """Result of export eligibility check.

    Story 8-4: AC #5 - Export blocked if < 70% confidence findings unverified.
    Story 12-3: AC #2 - Warnings for 70-90% confidence unverified findings.
    Story 3.2: Includes verification_mode for UI display.
    """

    model_config = ConfigDict(populate_by_name=True)

    eligible: bool = Field(..., description="True if export is allowed")
    verification_mode: str = Field(
        default="advisory",
        alias="verificationMode",
        description="Matter's verification mode: 'advisory' or 'required' (Story 3.2)",
    )
    blocking_findings: list[ExportBlockingFinding] = Field(
        default_factory=list,
        alias="blockingFindings",
        description="Findings blocking export (< 70% confidence, unverified in advisory; all pending in required)",
    )
    blocking_count: int = Field(0, ge=0, alias="blockingCount", description="Number of blocking findings")
    warning_findings: list[ExportWarningFinding] = Field(
        default_factory=list,
        alias="warningFindings",
        description="Findings with warnings (70-90% confidence, unverified) - empty in required mode",
    )
    warning_count: int = Field(0, ge=0, alias="warningCount", description="Number of warning findings")
    message: str = Field(
        "",
        description="Human-readable eligibility message",
    )
